lib/axon.ex

```diff
@@ -3139,93 +3139,327 @@ defmodule Axon do
     %{axon | hooks: [{on_event, mode, fun} | hooks]}
   end
 
-  ## Traversal
+  ## Graph Manipulation and Utilities
 
   @doc """
-  Traverses a model tree applying `fun` to each layer.
+  Returns a node's immediate parent(s).
+
+  You can use this to slice off the heads of pre-trained
+  models, or to manipulate nodes other graph manipulation
+  APIs.
   """
-  def tree_map(%Axon{op: :container, parent: [container]} = axon, fun) do
-    x = deep_new(container, fun)
-    %{fun.(axon) | parent: [x]}
-  end
+  def get_parent(%Axon{parent: [parent]}), do: parent
 
-  def tree_map(%Axon{parent: x} = axon, fun) when is_list(x) do
-    x = Enum.map(x, &tree_map(&1, fun))
-    %{fun.(axon) | parent: x}
-  end
+  def get_parent(%Axon{op: :container, parent: parent}), do: parent
+
+  def get_parent(%Axon{parent: parents}), do: parents
 
   @doc """
-  Traverses a model applying `fun` with an accumulator.
+  Sets a node's parent to the given Axon graph(s).
+
+  The new parents must be compatible with the original
+  parents of the given node. Note that replacing a node's
+  parent replaces it's entire lineage.
   """
-  def tree_reduce(%Axon{op: :container, parent: [container]} = axon, acc, fun) do
-    deep_reduce(container, fun.(axon, acc), fun)
-  end
+  def set_parent(%Axon{parent: [parent]} = axon, %Axon{output_shape: shape} = new_parent) do
+    %Axon{output_shape: parent_shape} = parent
+    # TODO: compatible? does not handle container shapes
+    unless Axon.Shape.compatible?(parent_shape, shape) do
+      raise ArgumentError,
+            "new parent graph must be compatible with old parent" <>
+              " graph, but got incompatible shapes #{inspect(parent_shape)}" <>
+              " and #{inspect(shape)}"
+    end
 
-  def tree_reduce(%Axon{parent: x} = axon, acc, fun) when is_list(x) do
```

Comment:
I guess I should probably deprecate this instead?

Reply:
Nah, it is fine. :) It is very early, best to move forward.